packages/framework/client-logger/external-telemetry/src/container/telemetryProducer.ts

```diff
@@ -0,0 +1,107 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	IContainer,
+	IContainerEvents,
+	type ICriticalContainerError,
+} from "@fluidframework/container-definitions";
+import {
+	ContainerConnectedTelemetry,
+	ContainerTelemetryEventNames,
+	IContainerTelemetry,
+	ContainerDisconnectedTelemetry,
+	ContainerClosedTelemetry,
+	ContainerAttachingTelemetry,
+	ContainerAttachedTelemetry,
+	type ContainerTelemetryEventName,
+} from "./containerTelemetry";
+import { ContainerSystemEventName, ContainerSystemEventNames } from "./containerSystemEvents";
+
+/**
+ * This class produces {@link IContainerTelemetry} from raw container system events {@link IContainerEvents}.
+ * The class contains different helper methods for simplifying and standardizing logic for adding additional information necessary
+ * to produce different {@link IContainerTelemetry}.
+ *
+ * @internal
+ */
+export class ContainerEventTelemetryProducer {
+	constructor(private container: IContainer) {}
+
+	public produceTelemetry(
+		eventName: ContainerSystemEventName,
+		payload?: any,
+	): IContainerTelemetry | undefined {
+		let telemetry: IContainerTelemetry | undefined = undefined;
+		switch (eventName) {
+			case ContainerSystemEventNames.CONNECTED:
+				telemetry = this.produceConnectedTelemetry(payload);
+				return telemetry;
+			case ContainerSystemEventNames.DISCONNECTED:
+				telemetry = <ContainerDisconnectedTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.DISCONNECTED)
+				);
+				break;
+			case ContainerSystemEventNames.CLOSED:
+				telemetry = this.produceClosedTelemetry(payload);
+				break;
+			case ContainerSystemEventNames.ATTACHED:
+				telemetry = <ContainerAttachedTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.ATTACHED)
+				);
+				break;
+			case ContainerSystemEventNames.ATTACHING:
+				telemetry = <ContainerAttachingTelemetry>(
+					this.produceBasicContainerTelemetry(ContainerTelemetryEventNames.ATTACHING)
+				);
+				break;
+			default:
+				break;
+		}
+		return telemetry;
+	}
+
+	private produceBasicContainerTelemetry = <T extends IContainerTelemetry>(
+		eventName: ContainerTelemetryEventName,
+	): T => {
```

Comment:
Doesn't look like this function needs to care about all the inherited types. As long as it returns an `IContainerTelemetry` everything is fine. That should also avoid the need of all the casts in `produceTelemetry()`.
```suggestion
	private produceBasicContainerTelemetry = (
		eventName: ContainerTelemetryEventName,
	): IContainerTelemetry => {
```

Reply:
Yeah I thought perhaps the casts would be useful but there isn't a need right now. Updated to just be IContainerTelemetry.